datafusion/functions/src/core/least.rs

```diff
@@ -0,0 +1,283 @@
+    if data_types.is_empty() {
+        plan_err!("least was called without any arguments. It requires at least 1.")
+    } else if let Some(coerced_type) = type_union_resolution(data_types) {
+        Ok(coerced_type)
+    } else {
+        plan_err!("Cannot find a common type for arguments")
+    }
+}
+
+impl ScalarUDFImpl for LeastFunc {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "least"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
+        Ok(arg_types[0].clone())
+    }
+
+    fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
+        if args.is_empty() {
+            return exec_err!(
+                "least was called with no arguments. It requires at least 1."
+            );
+        }
+
+        // Some engines (e.g. SQL Server) allow least with single arg, it's a noop
+        if args.len() == 1 {
+            return Ok(args[0].clone());
+        }
+
+        // Split to scalars and arrays for later optimization (constant folding)
+        let (scalars, arrays): (Vec<_>, Vec<_>) = args.iter().partition(|x| match x {
+            ColumnarValue::Scalar(_) => true,
+            ColumnarValue::Array(_) => false,
+        });
+
+        let mut arrays_iter = arrays.iter().map(|x| match x {
+            ColumnarValue::Array(a) => a,
+            _ => unreachable!(),
+        });
+
+        let first_array = arrays_iter.next();
+
+        let mut smallest: ArrayRef;
+
+        // Optimization: merge all scalars into one to avoid recomputing (constant folding)
+        if !scalars.is_empty() {
+            let mut scalars_iter = scalars.iter().map(|x| match x {
+                ColumnarValue::Scalar(s) => s,
+                _ => unreachable!(),
+            });
+
+            // We have at least one scalar
+            let mut smallest_scalar = scalars_iter.next().unwrap();
+
+            for scalar in scalars_iter {
+                smallest_scalar = keep_smaller_scalar(smallest_scalar, scalar)?;
+            }
+
+            // If we only have scalars, return the smaller one
+            if arrays.is_empty() {
+                return Ok(ColumnarValue::Scalar(smallest_scalar.clone()));
+            }
+
+            // We have at least one array
+            let first_array = first_array.unwrap();
+
+            // Start with the smaller value
+            smallest = keep_smallest(
+                Arc::clone(first_array),
+                smallest_scalar.to_array_of_size(first_array.len())?,
+            )?;
+        } else {
+            // If we only have arrays, start with the first array
+            // (We must have at least one array)
+            smallest = Arc::clone(first_array.unwrap());
+        }
+
+        for array in arrays_iter {
+            smallest = keep_smallest(Arc::clone(array), smallest)?;
+        }
```

Comment:
@waynexia this is the same as the greatest, see why https://github.com/apache/datafusion/pull/12474#discussion_r1885347287